### ISSUE TITLE
hepdata DOIs not resolving

### ISSUE THREAD

**Issue by tsgit** [HEPData/hepdata#120]
Hi,

how long is the delay between HepData minting a DOI and it resolving on dx.doi.org?

When I look at https://www.hepdata.net/search/?q= the first few DOIs do not resolve, and each of them comes with dozens of tables.

http://dx.doi.org/10.17182/hepdata.78924
http://dx.doi.org/10.17182/hepdata.78924.v2/t1
...

Consequently, when I look at 
http://inspirehep.net/search?cc=Data&ln=en&jrec=1
the first ca 500 DOIs do not resolve 

How long (in days) is the delay between a DOI being minted and it resolving? Do we need to adjust for that in the UI?

Cheers
T.

**Comment by GraemeWatt** [comment]
Dear Thorsten,

Thanks for pointing out this problem.  There should be no significant delay between a DOI being minted and it being resolved.  Looking at the log files I see that the DOI minting failed for some fraction of records with an error message:
```
File "/opt/hepdata/src/hepdata/hepdata/modules/records/utils/doi_minter.py", line 103, in create_container_doi
    base_xml, publication_info['uuid'])
KeyError: 'uuid'
```
On the other hand the DOI minting was successful for some recent records like https://www.hepdata.net/record/ins1495243 (the last one added).  I'll try to work out why the `uuid` key is missing for some records.  Debugging the DOI minting code locally is a bit difficult without the DataCite credentials.  We are also currently missing the functionality to mint DOIs after a record has been finalised (if the DOI minting failed the first time).  I think @eamonnmag started working on that.  We would then need to identify all records where the DOI minting failed and try again.  This might take some time to sort out, so if you need stable links from INSPIRE in the short term, it might be better just to use the HEPData URLs rather than the DOIs.  Apologies for the inconvenience.

Best regards,
Graeme Watt (Database Manager for HEPData)

**Comment by eamonnmag** [comment]
Yes, needless to say, still things to finish. The uuid thing is a mystery, not sure if something happened with some invenio dependency at some point. I have started the script to check and re-issue dois, but need some time to finish and test.

**Comment by GraemeWatt** [comment]
I looked into this issue today.  The DOI minting is successful if an Inspire ID is attached to a HEPData record when it is created, but *not* if the Inspire ID is attached at a later stage in the submission process, due to these lines:
https://github.com/HEPData/hepdata/blob/2c1ab6ee6fbcd1031773a7edbb749b1ace8ae5c9/hepdata/modules/records/views.py#L614-L615
These lines remove keys from `record` (including `uuid`!) that are not present in `content`.  This is fairly easy to fix, but then we need to go back to existing records to add the missing `uuid` keys and repeat the DOI minting.  I'll try to write some code that does this.

**Comment by GraemeWatt** [comment]
I've addressed this issue in da0dbcb4c064ec984a4461b58d033f9871105971 (and minor follow-ups b5fda0318e59147873119462f25f5481215e42e5 and 599e6a1c57e3eead8be13e668858ba1f00f78162), and run the scripts on the server to add the missing `uuid` keys and mint the DOIs.  There were about 25 HEPData records affected by this issue.

I also found another issue where the DOI minting failed due to Unicode characters in the data abstract.  I fixed the code and minted DOIs for a couple of records which had this problem, but there might be others.

In future, we should do a more systematic check that all HEPData records have resolved DOIs (and fix the problem with duplicates that @eamonnmag was investigating), but I'll close this issue for now.
